Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,255 @@
+                + " pointing to a file defining the credentials. See {1} for more information.",
+                CREDENTIAL_ENV_VAR,
+                HELP_PERMALINK));
+        }
+
+        /// <summary>
+        /// Unsynchronized helper method to obtain the Application Default Credentials.
+        /// </summary>
+        private ICredential GetDefaultCredentialUnsynchronized()
+        {
+            string credentialPath = null;
+            
+            // 1. First try the environment variable
+            try
+            {
+                credentialPath = GetEnvironmentVariable(CREDENTIAL_ENV_VAR);
+                if (!String.IsNullOrWhiteSpace(credentialPath))
+                {
+                    return LoadCredentialFromFile(credentialPath);
+                }
+            }
+            catch (Exception e)
+            {
+                // Catching generic exception type because any corrupted file could manifest in different ways including 
+                // but not limited to the System.IO or from the Newtonsoft.Json namespace
+
+                throw new InvalidOperationException(String.Format("Error reading credential file from location {0}: {1}"
+                                        + "\nPlease check the value of the Environment Variable {2}",
+                                        credentialPath,
+                                        e.Message,
+                                        CREDENTIAL_ENV_VAR));
+            }
+
+            // 2. Then try the well known file
+            try
+            {
+                credentialPath = GetWellKnownCredentialFilePath();
+                if (!String.IsNullOrWhiteSpace(credentialPath))
+                {
+                    return LoadCredentialFromFile(credentialPath);
+                }
+            }
+            catch (FileNotFoundException)
+            {
+                // File is not present, eat the exception and move on to the next check.
+            }
+            catch (IOException e)
+            {
+                // Catching generic exception type because any corrupted file could manifest in different ways including 
+                // but not limited to the System.IO or from the Newtonsoft.Json namespace
+
+                throw new InvalidOperationException(String.Format("Error reading credential file from location {0}: {1}"
+                                        + "\nPlease rerun 'gcloud auth login' to regenerate credentials file.",
+                                        credentialPath,
+                                        e.Message));
+            }
+
+            // 3. Then try the compute engine, if not already checked         
+            if (!checkedComputeEngine)
+            {
+                Logger.Info("Checking whether the application is running on ComputeEngine.");
+                bool isRunningOnComputeEngine = ComputeCredential.IsRunningOnComputeEngine().Result;
```

Comment:
We don't want to expose this as async APIs to the end consumer. Also, we want to leave the IsRunningOnComputeEngine() as async as this could be consumed elsewhere in future. Hence why we are terminating the async at this layer with the .Result.